### PR TITLE
Make loadGltfUris use promises

### PR DESCRIPTION
**Pull request by lasalvavida**
Part of #99.

@lilleyse can you look at this when you get a chance?

### CODE REVIEW

File: lib/loadGltfUris.js [hunk 1]
```diff
@@ -1,134 +1,99 @@
...
+function loadGltfUris(gltf, options) {
+    return new Promise(function(resolve, reject) {
+        var basePath = options.basePath;
+        var loadURIs = [
+            loadURI(gltf, basePath, 'buffers'),
+            loadURI(gltf, basePath, 'images'),
+            loadURI(gltf, basePath, 'shaders')
+        ];
+        Promise.all(loadURIs).then(function() {
+            gltf.extras._pipeline.jimpScratch = new Jimp(1, 1, function(err) {
```
Comment (lilleyse) [inline lib/loadGltfUris.js]: Hm, I haven't looked into this at all but is it possible to promisify a constructor's callback?

File: lib/loadGltfUris.js [hunk 2]
```diff
@@ -147,20 +112,48 @@ function isTransparent(image) {
     return false;
 }
 
-//Generate a jimp image for png, jpeg, or bmp
-function generateJimpImage(object, callback) {
+function readFromFile(object, name, uriPath) {
+    return new Promise(function(resolve, reject) {
+        fs.readFile(uriPath, function (err, data) {
```
Comment (lilleyse) [inline lib/loadGltfUris.js]: Promisify this function like in the node guide.

File: lib/loadGltfUris.js [hunk 3]
```diff
@@ -147,20 +112,48 @@ function isTransparent(image) {
...
-        if (jimpErr) {
-            throw jimpErr;
-        }
-
-        pipelineExtras.jimpImage = image;
-        pipelineExtras.imageChanged = false;
-        pipelineExtras.transparent = isTransparent(image);
-        callback();
+    return new Promise(function(resolve, reject) {
+        Jimp.read(pipelineExtras.source, function (err, image) {
```
Comment (lilleyse) [inline lib/loadGltfUris.js]: Same here.

File: specs/lib/loadImageUrisSpec.js [hunk 4]
```diff
@@ -34,7 +34,7 @@ describe('loadImageUris', function() {
         };
 
         addPipelineExtras(gltf);
-        loadGltfUris(gltf, options, function(err, gltf) {
+        loadGltfUris(gltf, options).then(function(f) {
```
Comment (lilleyse) [inline specs/lib/loadImageUrisSpec.js]: Remove the `f`

File: lib/loadGltfUris.js [hunk 5]
```diff
@@ -1,134 +1,99 @@
...
         }
     }
+    return Promise.all(promises);
 }
 
-//Return the extension of the data uri
+/**
+ * Return the extension of the data uri
+ * @param {string} uri
+ * @returns {string}
```
Comment (lilleyse) [inline lib/loadGltfUris.js]: Usually the type is capitalized.

**Comment by lilleyse** [comment]
This looks good so far, I can already tell promises are going to clean up a lot of areas.

File: lib/loadGltfUris.js (hunk 1 above)
Comment (lasalvavida) [inline lib/loadGltfUris.js]: Looking at the Jimp doc, this callback seems unnecessary

**Comment by lasalvavida** [comment]
@lilleyse Updated

**Comment by lilleyse** [comment]
Sorry for not pointing this out before, but in the node guide the promises code is styled like:

```javascript
compressFile('myFile')
    .then(function (data) {
        //Do something with data.
    })
    .catch(function(err){
        console.log(err);
    });
```

Which is different than the formatting here. @JoshuaStorm also take note here in case you are formatting your promises similarly.

**Comment by lasalvavida** [comment]
@lilleyse Updated with the formatting changes

**Comment by lilleyse** [comment]
Some of the `.catch` aren't on their own lines

**Comment by lasalvavida** [comment]
> Some of the .catch aren't on their own lines

Yup, you're right, sorry about that. I think I got them all this time.

**Comment by lilleyse** [comment]
Looks good!

**Pull request merged**